### ISSUE TITLE
What is the best way to store a Key in a property similar to { type: Key } ? 

### ISSUE THREAD

**Issue by dpsifr** [sebelga/gstore-node#130]

**Comment by dpsifr** [comment]
The data I have was created with Python/NDB and I am switching over to nodejs with this (really great!) library. When retrieving the existing data containing a key, it looks like this:

event_selected: {
  id: "6226140502949888",
  kind: "Event",
  path: [
    "Event",
    "6226140502949888"
  ]
}

So, it should be fairly easy to write a custom function to grab that object when returning the list and insert it in. And then to convert a key into this format when inserting.

**Comment by sebelga** [comment]
Actually yes it would be fearly easy. (until I have time to add a proper way to do it! :smile: )

You might have something like (not tested)

```js
const gstore = require('gstore-node');

MyEntity.list()
    .then(({ entities }) => {
        return Promise.all(...entities.map((entity) => (
            gstore.model('Event')
                .get(entity.event_selected.id)
                .then((eventSelected) => {
                    entity.event_selected = eventSelected;
                    return entity;
                })
        )))
        .then((entities) => {
            // ...
        })
    })
```

**Comment by dpsifr** [comment]
Thanks! I will give that a try and see about a generic implementation. I use keys pretty extensively across our data.

**Comment by dpsifr** [comment]
I'm getting an error...

What event is not being handled?

Error: Schema Event missing
    at Gstore.model index.js:71:23
    at User.findOne.then api.js:29:14
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

**Comment by dpsifr** [comment]
Ended up having to do: 

entity.entityData.event_selected = eventSelected.entityData

**Comment by dpsifr** [comment]
This is a little ugly but working (couldn't get map to work with some objects)...

`const resolveKeys = function (entity) {
  let promise = new Promise(function (resolve) {
    Object.keys(entity.entityData).forEach(function (key) {
      let keyValue = entity.entityData[key]
      if (Datastore.isKey(keyValue)) {
        gstore.model(keyValue.kind)
          .get(keyValue.id)
          .then((childEntity) => {
            entity.entityData[key] = childEntity.entityData
            resolve(entity)
          })
          .catch(function () {
            resolve(null)
          })
      }
    })
  })
  return promise
}`

**Comment by sebelga** [comment]
Be careful that it looks like you might never resolve the Promise if there is no _Key_ prop.

Also you might be interested in providing a[ Dataloader instance](https://sebelga.gitbooks.io/gstore-node/content/dataloader.html#dataloader) to this helper function and then forward it to the `gstore.model(...).get(...)` call.

I will put this feature as high priority in for the next release.

**Comment by dpsifr** [comment]
Definitely, I appreciate the priority. Your library is a life saver on huge project we are working on launching! 
I am taking a look at Dataloader today.

**Comment by dpsifr** [comment]
I like this a lot, thank you! I will start switching over. In the meantime,
I've had to do some workarounds and deal with one issue that I am not sure
how populate will handle.
When you get an entity with populate and want to modify something and save
it back. Does the save/update deal with going back to keys so it doesn't
save the populated data back into the entity.

**Comment by dpsifr** [comment]
Also, to facilitate this for multiple results, how do I specify an array of
type keys on a schema property.

**Comment by sebelga** [comment]
So both question would be an improvement on the current implementation 😊 

For the first one, unfortunately gstore does not automatically puts back the Key before saving.

You might want to have a helper for that purpose that might look like

```js
const savePopulatedEntity = (entity, refs = []) => {
  const data = {};

  refs.forEach(ref => {
    // Save the data
    data[ref] = entity.entityData[ref];

    // Put back the Key
    entity.entityData[ref] = data[ref][gstore.ds.KEY];
  });

  return entity.save().then((updatedEntity) => {
    // Put back the entityData
    refs.forEach(ref => updatedEntity.entityData[ref] = data[ref]);
    return updatedEntity;
  });
};

// That you can call like that
savePopulatedEntity(user, ['address', 'company'];
```

For the second feature, it needs to be added 😊

**Comment by dpsifr** [comment]
Cool, I appreciate the quick response and the new implementation.
I do really like your approach and I'm going through a lot of code now to
utilize it!

**Comment by dpsifr** [comment]
I've also done a helper function similar to this that includes code to add the ID to elements for serializing to the front end and then removing it before saving/updated. It became necessary because of the change to Google's implementation on keys and how they return them. I wasn't sure if there was a better way to deal since I do need an id on the element for the front end to know what it was working on.

I'm wondering if it would make sense for populate to add this on populate() and remove on save. It's similar to how the python implementation worked.

**Comment by sebelga** [comment]
I will think about the best solution when saving back the entity. In the meantime the `populate()` will add an `id` property of the reference entity fetched.

**Comment by dpsifr** [comment]
That's perfect. I know plain() does this, but now when you are resolving
like in populate() or on multiple returned entities.

**Comment by dpsifr** [comment]
What if the SavePopulatedEntity function was just part of the validate() portion of the flow. Then you could call anything with populate() to resolve everything and when it is validated before save, anything with type of Key would be turned back into a Key before saving. Seamless two way serialization and maybe promoting better datastore design ;)

**Comment by sebelga** [comment]
Yes that might work, I am looking into it. 👍

**Comment by dpsifr** [comment]
It will require having keys on each populate()ed child which I didn't see currently. I'm excited about this implementation. Going to help a lot.

**Comment by sebelga** [comment]
Yes each populated entity has the Key in `gstore.ds[KEY]`. I am thinking in maybe allows saving all referenced entities if needed. Something like `entity.save(null, { saveAllRefs: true })`. That would allow to make change to any of the populated entities and save them all in one call.

The problem is, it will replace **everything** in the Datastore, so if a ref entity has a `read: false` property, it won't be there when populating and so would be lost.

Giving it some thoughts...

**Comment by dpsifr** [comment]
Anyway to support the showkey parameter on the populated children? It's a
consistency thing when dealing with a result  and serializing to the front
end to reference the ids.

**Comment by dpsifr** [comment]
I don't see the gstore.ds[KEY], I've tried a few way to access and don't see in debugger....
I resolve the user object with populate() that populates the eventSelected property and need the id or key on that property...

None work:
req.session.user.entityData.eventSelected[gstore.ds.KEY]
req.session.user.entityData.eventSelected.gstore.ds.KEY
req.session.user.entityData.eventSelected.entityKey
req.session.user.entityData.eventSelected.__key

No luck. I'm sure I'm just being dense here because I'm buried in a few hundred other issues.

**Comment by sebelga** [comment]
The `id` you should have it. And for the `gstore.ds.KEY` are you sure you have the **same instance** of gstore in both files?

Here is what you should have if you put a breakpoint after populating

<img width="801" alt="screen shot 2019-02-08 at 08 15 10" src="https://user-images.githubusercontent.com/2854616/52464184-36fc0000-2b7a-11e9-9011-5bc79e3d2efb.png">

In the "user" populated, you can see the `id: "1"` and the Symbol KEY

**Comment by sebelga** [comment]
I will close this issue as it has been resolved. Feel free to open another one if you have another issue. Thanks!

**Comment by sebelga** [comment]
Hello,
FYI I just released `v5.0.2` that fix the problem to save an entity that was previously populated.
Cheers
